osdep/timer-win2.c

```diff
@@ -19,22 +19,51 @@
 #include <sys/time.h>
 #include <mmsystem.h>
 #include <stdlib.h>
+#include <versionhelpers.h>
+
 #include "timer.h"
 
 #include "config.h"
 
 static LARGE_INTEGER perf_freq;
 
+// ms values
+static int hires_max = 50;
+static int hires_res = 1;
+
+int mp_start_hires_timers(double wait_ms)
+{
+#if !HAVE_UWP
+    // policy: request hires_res ms resolution if wait < hires_max ms
+    if (wait_ms > 0 && wait_ms < hires_max &&
+        timeBeginPeriod(hires_res) == TIMERR_NOERROR)
+    {
+        return hires_res;
+    }
+#endif
+    return 0;
+}
+
+void mp_end_hires_timers(int res_ms)
+{
+#if !HAVE_UWP
+    if (res_ms > 0)
+        timeEndPeriod(res_ms);
+#endif
+}
+
 void mp_sleep_us(int64_t us)
 {
     if (us < 0)
         return;
+    int hrt = mp_start_hires_timers(us / 1e3);
```

Comment:
> Otherwise <1000μs waits won't change the timer freq, even though `Sleep(1)` will be called.

They would change the resolution, because 1e3 is double, hence the value would be >0. Nevertherless I'll do the change you request, but it's no-op.

Reply:
Right, I didn't realise ``mp_start_hires_timers`` had a double argument. I think it should be an int, since timeBeginPeriod, Sleep and SleepConditionVariable* all take an integer argument.

Reply:
> Right, I didn't realise `mp_start_hires_timers` had a double argument. I think it should be an int, since timeBeginPeriod, Sleep and SleepConditionVariable* all take an integer argument.

I intentionally made it a double so that any value bigger than 0 (in any timeout resolution, including sub-ms or even sub-us) would trigger hires. This is pretty much in line with the logic at `mp_sleep_us` for sub-ms values (and the reason why the change you requested initially is almost no-op), and I still think it's the correct thing to do.

Agreed?

Reply:
Well, to be in line with the mpv functions, it would have to take int64 microseconds. As a double it's not in line with Windows conventions or mpv conventions. Since we have to convert the timeout to milliseconds for Windows anyway (including rounding up the timeout value to 1ms,) I think it makes sense to pass the exact same integer value to ``mp_start_hires_timers`` that we pass to the Windows API.

Reply:
> to be in line with the mpv functions, it would have to take int64 microseconds. As a double it's not in line with Windows conventions or mpv conventions

I still think double is the most generic and native way to express a value which can then be used with logic of "anything which isn't exactly 0 should get hires timers" regardless of platform or application conventions, however, I also don't think it's worth spending more time on this, and at least currently, using microsec argument doesn't reduce its usability.

I'll prefer it int64 microsec value to be aligned with the mpv conventions. I don't think it should align with the windows conventions, because the timeout value comes from mpv, and that's what this function should take into account in this context - regardless of the value which ends up being used with the windows API.

Are you OK with int64 microsec?

Reply:
I'll admit that part of this is because I don't like unnecessary use of floating point arithmetic (even though I know there aren't accuracy problems in this case) so I'm okay with int64 microseconds.

I'm not sure why you're against integer milliseconds though. Both call sites have an integer millisecond value easily available, and the other values that ``mp_start_hires_timers`` uses (``hires_max`` and ``hires_res``) are also integer milliseconds, so there would be less conversion involved if ``wait_ms`` was integer milliseconds too. ``mp_start_hires_timers`` isn't a public API, so we could always change it to use microseconds if we need to in future.

Reply:
> I'm not sure why you're against integer milliseconds though

Because it can't represent the concept which the logic applies - anything which isn't exactly 0, e.g. even one nanosec.

If 0 was included in the range of hires, then it wouldn't matter to me.

Reply:
Changed to int.